rmw_fastrtps_shared_cpp/src/custom_publisher_info.cpp

```diff
@@ -96,6 +96,46 @@ bool PubListener::take_event(
         incompatible_qos_status_.total_count_change = 0;
       }
       break;
+    case RMW_EVENT_PUBLICATION_MATCHED:
+      {
+        auto rmw_data = static_cast<rmw_matched_status_t *>(event_info);
+        if (matched_changes_) {
+          rmw_data->total_count_change = matched_status_.total_count_change;
+          rmw_data->current_count = matched_status_.current_count;
+          rmw_data->current_count_change = matched_status_.current_count_change;
+
+          matched_changes_ = false;
+        } else {
+          eprosima::fastdds::dds::PublicationMatchedStatus matched_status;
+          publisher_info_->data_writer_->get_publication_matched_status(matched_status);
+          rmw_data->total_count_change = matched_status.total_count_change;
+          rmw_data->current_count = matched_status.current_count;
+          rmw_data->current_count_change = matched_status.current_count_change;
+        }
+        matched_status_.current_count_change = 0;
+        matched_status_.total_count_change = 0;
```

Comment:
why is the if-else needed?
`get_publication_matched_status()` and the `on_publication_matched()` listener should reset the internal `current_count_change`/`total_count_change` counters.

As we are using the listener, we should use:

```suggestion
          std::lock_guard guard{on_new_event_m_};
          rmw_data->total_count_change = matched_status_.total_count_change;
          rmw_data->current_count = matched_status_.current_count;
          rmw_data->current_count_change = matched_status_.current_count_change;
          matched_status_.current_count = 0;
          matched_status_.current_count_change = 0;
```

Reply:
> why is the if-else needed?

I refer to the process of other event.  I don't sure which case may have no listener (Have to call get_publication_matched_status() to get status).   
If listener is always used, we need not get status by get_publication_matched_status(). I will update it.